### PR TITLE
Store component

### LINKED ISSUE #229: add localstorage/sessionstorage  support
It would be better that dash core components has localstorage or sessionstorage support,
The code is simple however useful.
And it is a better way to share state than hidden div from my view.

here is a quick experiment code:
https://github.com/Benjamin-Shengming/local_storage

### PR DESCRIPTION
**Pull request by T4rk1n**
**Available in dash-core-components==0.30.0rc1**

`pip install dash-core-components==0.30.0rc1`

# Store component

Alternative to the hidden div method for caching data on the front-end, no need to `json.dumps` and `json.loads`.

Three type of storage (`storage_type` prop):

- `memory` - Data is only kept in memory as prop by the component, cleared on refresh.
- `local` - Data is kept in `window.localStorage`, only cleared manually.
- `session` - Data is kept in `window.sessionStorage`, cleared on browser exit.

__** Data must be json serializable. **__

Example: 

```python
import dash

import dash_html_components as html
import dash_core_components as dcc
from dash.dependencies import Output, Input, State
from dash.exceptions import PreventUpdate

app = dash.Dash(__name__)

app.layout = html.Div([
    dcc.Store(id='storage'),
    html.Button('click me', id='btn'),
    html.Button('clear me', id='clear-btn'),
])

@app.callback(Output('storage', 'data'),
              [Input('btn', 'n_clicks')],
              [State('storage', 'data')])
def on_click(n_clicks, storage):
    if n_clicks is None:
        raise PreventUpdate
    storage = storage or {}
    return {'clicked': storage.get('clicked', 0) + 1}


@app.callback(Output('storage', 'clear_data'),
              [Input('clear-btn', 'n_clicks')])
def on_clear(n_clicks):
    if n_clicks is None:
        raise PreventUpdate
    return True


if __name__ == '__main__':
    app.run_server(debug=True, threaded=True, port=11111)
```

cc @plotly/dash 

Closes #229

### CODE REVIEW

**Comment by bpostlethwaite** [comment]
cc @nicolaskruchten - I know you have were thinking about a dcc.Store component.

**Comment by nicolaskruchten** [comment]
Yep, although I was mostly concerned about security: https://github.com/plotly/dash/issues/262

**Comment by mkhorton** [comment]
I think this is a great idea, even without server-side encryption this would be a nice improvement over the current hidden div hack.

**Comment by mkhorton** [comment]
Also, this seems orthogonal to the issue of whether or not to use e.g. Redis too.

A more comprehensive solution using something like Redis to store user data would still be useful if there's large user-specific data that only needs to exist server-side, since it would reduce traffic needed to pass data between components, but that obviously requires a lot of extra set-up. Being able to use HTML localStorage with 'off-the-shelf' Dash would be really nice in my opinion, and be sufficient for most people (myself included).

**Comment by rmarren1** [comment]
This seems a lot cleaner than the hidden div hack, and I think it is useful enough to release without encryption and then add encryption as a separate PR, as long as we make it clear in the docs the stored data is available to the client

**Comment by chriddyp** [comment]
This looks good to me.

My main feedback is about the name. I wonder if we should be clear that this is client-side storage. Some different name ideas:
- `Store`
- `BrowserStore`
- `ClientStore`
- `Storage`
- `BrowserStorage`
- `ClientStorage`

I think I prefer `BrowserStore`

I would also like @ned2 's opinion on this.

Otherwise, I'm 💃 on this in principle. I haven't taken a look into the code

**Comment by chriddyp** [comment]
> local - Data is kept in window.localStorage, only cleared manually.
> session - Data is kept in window.sessionStorage, cleared on browser exit.

I'd also like to play around with this a little bit actually. It's not clear to me how this would fit into a set of callbacks.

i.e. how would you structure your code so that the backend knows whether data is already available in the store or it needs to be added?

perhaps we could add a lightweight "`date_modified`" property and the dev could have that property as an `Input` to their callback. If it's `-1`, then they need to supply data. Otherwise, they can raise `PreventUpdate`:
```
@app.callback(Output('datastore', 'value'), [Input('datastore', 'date_modified')])
def update_datastore(date_modified):
    if date_modified is None or date_modified == -1:
        raise dash.exceptions.PreventDefault()
    return expensive_computation()
```

File: src/components/Storage.react.js [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+ * The data is not inserted in the DOM.
+ * Data can be in memory, localStorage or sessionStorage.
+ * The data will be kept with the id as key.
+ */
+export default class Storage extends React.Component {
+    constructor(props) {
+        super(props);
+
+        if (props.storage_type === 'local') {
+            this._backstore = _localStore;
```
[the remark this answers is not in the document]
Comment (T4rk1n) [reply src/components/Storage.react.js]: It also wrap `getItem` with `JSON.parse` and and `setItem` with `JSON.stringify`.

**Comment by T4rk1n** [comment]
@plotly/dash Ready for another round of reviews, I added a timestamp for when the data is modified. 

Available in version 0.30.0rc1.

I made a poll for the component name: https://strawpoll.com/ed9ychp8

**Comment by T4rk1n** [comment]
Renaming the storage component to `Store` according to the poll results.

```
54.55 % (6 votes)Store
18.18 % (2 votes)ClientStore
18.18 % (2 votes)Storage
9.09 % (1 votes)BrowserStore
0 % (0 votes)BrowserStorage
0 % (0 votes)ClientStorage
11 total votes.
```

**Comment by tguptaMT** [comment]
dash-core-components==0.30.0rc1 seems to hide my graph's axis labels somehow. If I switch to the latest version of dcc, the axis titles (labels) are visible but the dcc.Storage component isn't. If I move to dcc==0.30.0rc1, Storage is available but the axis labels on my graph disappear. Are there any syntax incompatibility issues with the following code:

`layout = go.Layout(
            xaxis=dict(
                title='Time (Weeks)',
                titlefont=dict(
                    family='Courier New, monospace',
                    size=18,
                    color='#07329C')),
            
            yaxis=dict(
                title='Median Aggresssion',
                titlefont=dict(
                    family='Courier New, monospace',
                    size=18,
                    color='#07329C')))`

**Comment by plopd** [comment]
Why is `dcc.Storage` not available in the latest version of dash-core-components==0.43.0

**Comment by lioneltrebuchon** [comment]
@plopd I think it was renamed to `dcc.Store` based on a previous poll.

**Pull request merged**